.circleci/initialize-manager.py

```diff
@@ -0,0 +1,52 @@
+#!/usr/bin/env python
+
+from fabric.api import execute
+
+from common import *
+# This is expected to be launch from the ci container
+from ci_variables import *
+
+def initialize_manager(max_runtime, filesystem_timeout):
+    """ Preforms the prerequisite tasks for all CI jobs that will run on the manager instance
+
+    max_runtime (seconds): The maximum uptime this manager should have be for it is terminated
+        This covers potential livelock scenario in one of the jobs (that produces fs activity)
+
+    filesystem_timeout (seconds):  If no fs activity is detected in ~/firesim after 2x this
+        interval, the manager is terminated.  CircleCI provides no mechanism for
+        doing cleanup after all jobs are done (even if some of those jobs fail).
+        So here we use FS activity as a proxy for doneness
+    """
+
+    # Catch any exception that occurs so that we can gracefully teardown the local jjjjjjjjjjjjjjjjjjjjjjj
+    try:
+        put(ci_workdir + "/scripts/machine-launch-script.sh", manager_home_dir)
+        with cd(manager_home_dir):
+            put(ci_workdir + "/scripts/machine-launch-script.sh", ".")
+            run("chmod +x machine-launch-script.sh")
+            run("sudo ./machine-launch-script.sh")
+            run("git clone https://github.com/firesim/firesim.git")
+
+        with cd(manager_fsim_dir):
+            run("git checkout " + ci_commit_sha1)
+            run("./build-setup.sh --fast")
+
+        with cd(manager_fsim_dir), prefix("source ./sourceme-f1-manager.sh"):
+            run(".circleci/firesim-managerinit.expect {} {} {}".format(
+                os.environ["AWS_ACCESS_KEY_ID"],
+                os.environ["AWS_SECRET_ACCESS_KEY"],
+                os.environ["AWS_DEFAULT_REGION"]))
+
+        with cd(manager_ci_dir):
+            # Set up a pair of crude checks to powerdown the instance
+            # The sleep 1 is required for the screen command to not be DOA
+            run("screen -d -m ./manager-watchdog.sh {} {} {}; sleep 1".format(max_runtime, filesystem_timeout, ci_workflow_id))
```

Comment:
I agree with your description assessment. This should probably be polling the workflow state to shut down in this current setup. 

With that being said, this might be problematic when running longer jobs (AFI creation, FPGA-sims). In that case, there is no "nice" way to run a long CI job (IIRC longer than 5hrs). For example, this prevents you from having a single job called `afi-creation` that lasts a long time. In that case, the BOOM CI creates separate workflows (spawned by the CircleCI API) to finish up when the AFI build passes/fails. This creates an issue where a single manager instance can have multiple workflows associated with it (assuming some version of what I did... spawning new workflows... is done here). Maybe the solution, in this case, is to associate one manager with multiple workflows and if all workflows are completed then the manager instance terminates everything on it and shuts down (note how "shutting down" is now a bit more complicated because if FPGA-sims are going... then you need to terminate those F1 instances). However, I think this should be fairly straight forward to do and much cleaner in Python than in Bash :).

Reply:
Actually on second thought... the following statement is wrong:

> Maybe the solution, in this case, is to associate one manager with multiple workflows and if all workflows are completed then the manager instance terminates everything on it

Technically, if we are using CircleCI to do this there might be periods where there are no workflows running but the AWS FireSim manager is running (building an AFI, running SPEC, etc). So there has to be some sort of tracking on the manager... just like what I did in the BOOM CI (have some file indicate what is running or isn't).

Reply:
Yeah, some manager side tracking will be required.

Although, I guess you could continuously spin up new polling jobs when they are about to reach their time limit... #hacks